kernel/src/hil/ble.rs

```diff
@@ -0,0 +1,247 @@
+//! BLE Upper and Lower level HILs
+//!
+//!
+//!       BLE Syscall Capsule
+//!
+//! +-------------------+  +--------------------+
+//! |   BleGattServer   |  |   BLEGattClient    |
+//! +-------------------+  +--------------------+
+//!
+//!
+//!        BLE Stack (eg: Rubble)
+//!
+//!
+//! +-------------------------------------------+
+//! |   BleRadio                                |
+//! +-------------------------------------------+
+//!
+
+use crate::ReturnCode;
+
+// *************************************************************************
+//   High Level HIL
+// *************************************************************************
+
+pub enum AttributeType {
+    Uuid16(u16),
+    Uuid128(u128),
+}
+
+pub const PRIMARY_SERVICE: AttributeType = AttributeType::Uuid16(0x2800);
+pub const SECINDARY_SERVICE: AttributeType = AttributeType::Uuid16(0x2801);
+pub const INCLUDE: AttributeType = AttributeType::Uuid16(0x2802);
+pub const CHARACTERISTIC: AttributeType = AttributeType::Uuid16(0x2802);
+
+/// Attribute format types from https://www.bluetooth.com/specifications/assigned-numbers/format-types/
+#[repr(u16)]
+pub enum AttributeFormatType {
+    Boolean = 0x01,
+    Bit2 = 0x02,
+    Nibble = 0x03,
+    Uint8 = 0x04,
+    Uint12 = 0x05,
+    Uint16 = 0x06,
+    Uint24 = 0x07,
+    Uint32 = 0x08,
+    Uint48 = 0x09,
+    Uint64 = 0x0a,
+    Uint128 = 0x0b,
+    Sint8 = 0x0c,
+    Sint12 = 0x0d,
+    Sint16 = 0x0e,
+    Sint24 = 0x0f,
+    Sint32 = 0x10,
+    Sint48 = 0x11,
+    Sint64 = 0x12,
+    Sint128 = 0x13,
+    Float32 = 0x14,
+    Float64 = 0x15,
+    SFloat = 0x16,
+    Float = 0x17,
+    DUint16 = 0x18,
+    UTF8s = 0x19,
+    UTF16S = 0x1a,
+    Struct = 0x1b,
+}
+
+pub enum BleGattAccessType {
+    None,
+    ReadOnly,
+    WriteOnly,
+    NotifyOnly,
+    ReadWrite,
+    ReadNotify,
+    WriteNotify,
+    ReadWriteNotify,
+}
+
+pub trait BleGattService<'a> {
+    type Characteristic: BleGattCharacteristic;
+
+    fn get_uuid(&self) -> AttributeType;
+    fn get_characteristics(&self) -> &'a [Self::Characteristic];
+}
+
+pub trait BleGattCharacteristic {
+    fn get_uuid() -> AttributeType;
+    fn get_format_type() -> AttributeFormatType;
+    fn get_access() -> BleGattAccessType;
+}
+
+pub enum BleGattError {
+    Timeout,
+}
+
+pub trait BleGattServer<'a> {
+    type Service: BleGattService<'a>;
+    /// Configure the GATT Server
+    ///
+    /// name - device name
+    /// services - a list of services
+    /// connect_interval_ms - interval (ms) to request the GATT Central to reconnect
+    fn configure(
+        &self,
+        name: &'a [u8],
+        services: &'a [Self::Service],
+        connect_interval_ms: u32,
+    ) -> ReturnCode;
+
+    // Disconnect GATT Central
+    fn disconnect(&self) -> ReturnCode;
+
+    // Start GAP Advertisement
+    fn start_advertisement(&self) -> ReturnCode;
```

Comment:
How does the advertisement contents get set?

Reply:
Using the `configure` function

Reply:
But there are more advertising parameters. We don't need to support them all initially necessarily, but having a function for future expansion would be helpful.

Reply:
I haven't seen anything else in the GAP advertising, but I'm not an expert in BLE. What parameters are you referring to?

Reply:
For example having manufacturer specific data in the advertisement, whether the device is connectable or not, long or short service UUIDs. There are a lot of advertisement options.